docs/dag-sharding-rfc.md

```diff
@@ -0,0 +1,157 @@
+
+
+### Adding files
+Currently the proposed implementation of basic sharding relies on ipfs-cluster having an incoming stream of data that can be read lazily as cluster sends off blocks to remote peers.  This allows sharding files that are too big to fit in a single repo or whose dags are too big to read into memory all at once, as resources can be cleared while streaming occurs.  Only the cids of blocks in the shard need to be maintained to construct the shard node of the cluster-dag, not the entire block or in-memory node structure.  Therefore in our basic sharding implementation we assume that sharded data must be added to ipfs-cluster directly.  This is because, as far as we know, there is not currently a way to access such a stream of blocks belonging to an already-imported-dag without potentially overloading the local ipfs daemon's resources.
+
+The cluster node adding the file will read in a stream of bytes that are then passed to an importer to create an ipfs files dag chunked (size-splitter or rabin-splitter) and layed out (trickle or balanced) per user instruction.  Before importing blocks of each shard the cluster peer calls the `allocate()` function to find the pinning peer for the upcoming shard.  Concurrent with ipfs files dag creation and transmission of the dag's blocks to the allocated cluster peer, the importing ipfs-cluster node will oversee construction of a cluster-dag shard node using the cids of the blocks of the ipfs files dag.  Once a shard is filled with cid references it will also be transmitted and pinned to the allocated ipfs-cluster node.  This process continues until both dags are fully constructed and pinned as expected.  The importing node transmits blocks by calling `ipfs block put` and `ipfs pin` on the pinning node's ipfs daemon via an RPC to the pinning cluster peer.   The pinning node should check the state to ensure it was actually assigned the pin before running the rpc to completion.  Note that this will be one of the more involved parts of implementing basic sharding and will involve extracting, refactoring and adding features to the dag building functionality of go-ipfs.
+
+There will likely be several endpoints entering into the functionality for adding sharded data.  For example, there will be a go endpoint `Add(r io.Reader, opts...)` which just adds arbitrary data. There will be an HTTP endpoint to which users can post multipart data just like it is now posted to ipfs and which creates the unixfs merkle dag.  If dags can be streamed node by node through ipfs safely without possibly overloading the process and its repo, cluster will likely support an AddDag(chan <-*dagNode) endpoint which allocates already-imported-dags.
```

Comment:
I would remove the last sentence. After the example I would add that how the API endpoints look will be defined in a later phase. But maybe that's a bit vague for a proposal. The idea is that:

* We'll be able to receive arbitrary data, chunk it and distribute it in the cluster
* We'll be able to receive filesystem data, chunk it and distribute it in the cluster
* We'll be able to receive a CID, and distribute the referenced tree in the cluster

Reply:
SGTM (also fixes github's italics formatting issue!)